monty-31/src/x86_64_avx2/packing.rs

```diff
@@ -160,21 +160,71 @@ fn add<MPAVX2: MontyParametersAVX2>(lhs: __m256i, rhs: __m256i) -> __m256i {
 // definition of Q and μ, we have Q P = μ C P = P^-1 C P = C (mod B). We also have
 // C - Q P = C (mod P), so thus D = C B^-1 (mod P).
 //
-// It remains to show that R is in the correct range. It suffices to show that -P <= D < P. We know
+// It remains to show that R is in the correct range. It suffices to show that -P < D < P. We know
 // that 0 <= C < P B and 0 <= Q P < P B. Then -P B < C - QP < P B and -P < D < P, as desired.
 //
 // [1] Modern Computer Arithmetic, Richard Brent and Paul Zimmermann, Cambridge University Press,
 //     2010, algorithm 2.7.
 
+// We provide 2 variants of Montgomery reduction depending on if the inputs are unsigned or signed.
+// The unsigned variant follows steps 1 and 2 in the above protocol to produce D in (-P, ..., P).
+// For the signed variant we assume -PB/2 < C < PB/2 and let Q := μ C mod B be the unique
+// representative in [-B/2, ..., B/2 - 1]. The division in step 2 is clearly still exact and
+// |C - Q P| <= |C| + |Q||P| < PB so D still lies in (-P, ..., P).
+
+/// Perform a Montgomery reduction on each 64 bit element.
+/// Input must lie in {0, ..., 2^32P}.
+/// The output will lie in {-P, ..., P} and be stored in the upper 32 bits.
+#[inline]
+#[must_use]
+#[allow(non_snake_case)]
+fn monty_red_unsigned<MPAVX2: MontyParametersAVX2>(input: __m256i) -> __m256i {
+    unsafe {
+        let q = x86_64::_mm256_mul_epu32(input, MPAVX2::PACKED_MU);
+        let q_P = x86_64::_mm256_mul_epu32(q, MPAVX2::PACKED_P);
+        x86_64::_mm256_sub_epi32(input, q_P)
+    }
+}
```

Comment:
Nit: There's no real benefit to the non-snake case of `q_P` (after all, `q` is `Q` in the docs above), so this would be cleaner as:
```suggestion
fn monty_red_unsigned<MPAVX2: MontyParametersAVX2>(input: __m256i) -> __m256i {
    unsafe {
        let q = x86_64::_mm256_mul_epu32(input, MPAVX2::PACKED_MU);
        let q_p = x86_64::_mm256_mul_epu32(q, MPAVX2::PACKED_P);
        x86_64::_mm256_sub_epi32(input, q_p)
    }
}
```